tests/model_registry/model_catalog/upgrade/test_catalog_upgrade.py

```diff
@@ -0,0 +1,371 @@
+
+        # Parse custom catalog entries properly
+        custom_catalog_entries = yaml.safe_load(f"catalogs:\n{custom_catalog_yaml}")["catalogs"]
+
+        # Check if custom catalog already exists to avoid duplicates
+        existing_catalog_ids = {catalog["id"] for catalog in existing_catalogs}
+        new_catalog_entries = [
+            catalog for catalog in custom_catalog_entries if catalog["id"] not in existing_catalog_ids
+        ]
+
+        # Combine catalogs properly as Python objects (only add if not already present)
+        if new_catalog_entries:
+            combined_catalogs = existing_catalogs + new_catalog_entries
+        else:
+            combined_catalogs = existing_catalogs
+            LOGGER.info(f"Custom catalog {TEST_DATA['catalog_id']} already exists, skipping addition")
+
+        combined_sources_yaml = yaml.dump({"catalogs": combined_catalogs}, default_flow_style=False)
+
+        # Store original configmap state for potential rollback
+        original_configmap_data = catalog_config_map.instance.data.copy()
+
+        # Prepare data for ResourceEditor pattern
+        new_data = {"sources.yaml": combined_sources_yaml, custom_sample_yaml_filename: custom_sample_yaml_content}
+
+        # Preserve any existing data that we're not modifying
+        for key, value in original_configmap_data.items():
+            if key not in new_data:
+                new_data[key] = value
+
+        # Apply permanent configuration changes using ResourceEditor.update() for persistent changes
+        patches = {catalog_config_map: {"data": new_data}}
+
+        ResourceEditor(patches=patches).update()
+
+        # Wait for model catalog to pick up changes
+        is_model_catalog_ready(client=admin_client, model_registry_namespace=model_registry_namespace)
+
+        # Verify Data Ingestion
+        @retry(wait_timeout=120, sleep=10, exceptions_dict={AssertionError: [], ResourceNotFoundError: []})
+        def verify_custom_catalog_ingestion():
+            sources_response = execute_get_command(
+                url=f"{model_catalog_rest_url[0]}sources", headers=model_registry_rest_headers
+            )
+
+            # Assert custom catalog exists alongside default catalog
+            custom_catalog_found = any(source["id"] == TEST_DATA["catalog_id"] for source in sources_response["items"])
+            assert custom_catalog_found, (
+                f"Custom catalog {TEST_DATA['catalog_id']} not found in sources: {[s['id'] for s in sources_response['items']]}"  # noqa: E501
+            )
+
+            # Verify default catalog still exists
+            default_catalog_found = any(source["id"] == DEFAULT_CATALOG_ID for source in sources_response["items"])
+            assert default_catalog_found, (
+                f"Default catalog should still exist alongside custom catalog. Available sources: {[s['id'] for s in sources_response['items']]}"  # noqa: E501
+            )
+
```

Comment:
minor: this could be extracted to a function with input params to avoid duplication

Reply:
this is already defined in a function